File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/handlers/StructuredMessages.java
package msdk.java.handlers;

import androidx.annotation.NonNull;

import com.evernym.sdk.vcx.connection.ConnectionApi;

import org.json.JSONObject;

import java9.util.concurrent.CompletableFuture;
import msdk.java.logger.Logger;

/**
 * Class containing methods to work with structured messages;
 */
public class StructuredMessages {
    private StructuredMessages() {

    }

    /**
     * answer structured message
     *
     * @param serializedConnection JSON string containing serialized connection
     * @param structuredMessage    structured message
     * @param answer               value of the answer
     * @return {@link CompletableFuture} containing message ID
     */
    public static @NonNull
    CompletableFuture<Void> answer(
            @NonNull String serializedConnection,
            @NonNull String structuredMessage,
            @NonNull JSONObject answer
    ) {
        Logger.getInstance().i("Respond to structured message");
        CompletableFuture<Void> result = new CompletableFuture<>();
        try {
            ConnectionApi.connectionDeserialize(serializedConnection).whenComplete((conHandle, err) -> {
                if (err != null) {
                    Logger.getInstance().e("Failed to deserialize connection: ", err);
                    result.completeExceptionally(err);
                    return;
                }
                try {
                    System.out.println(answer + "CompletableFuture1");
                    ConnectionApi.connectionSendAnswer(conHandle, structuredMessage, answer.toString()).whenComplete((res, er) -> {
                        if (er != null) {
                            Logger.getInstance().e("Failed to send answer: ", er);
                            result.completeExceptionally(er);
                        }
                        System.out.println(answer + "CompletableFuture2");
                        result.complete(null);
                    });
                } catch (Exception ex) {
                    result.completeExceptionally(ex);
                }
            });
        } catch (Exception ex) {
            result.completeExceptionally(ex);
        }
        return result;
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/handlers/StructuredMessages.kt
package msdk.kotlin.sample.handlers

import com.evernym.sdk.vcx.connection.ConnectionApi
import java9.util.concurrent.CompletableFuture
import msdk.kotlin.sample.logger.Logger
import org.json.JSONObject

/**
 * Class containing methods to work with structured messages;
 */
object StructuredMessages {
    /**
     * answer structured message
     *
     * @param serializedConnection JSON string containing serialized connection
     * @param structuredMessage    structured message
     * @param answer               value of the answer
     * @return [CompletableFuture] containing message ID
     */
    fun answer(
        serializedConnection: String,
        structuredMessage: String,
        answer: JSONObject
    ): CompletableFuture<Void?> {
        Logger.instance.i("Respond to structured message")
        val result =
            CompletableFuture<Void?>()
        try {
            ConnectionApi.connectionDeserialize(serializedConnection)
                .whenComplete { conHandle: Int?, err: Throwable? ->
                    if (err != null) {
                        Logger.instance
                            .e("Failed to deserialize connection: ", err)
                        result.completeExceptionally(err)
                        return@whenComplete
                    }
                    try {
                        println(answer.toString() + "CompletableFuture1")
                        ConnectionApi.connectionSendAnswer(
                            conHandle!!,
                            structuredMessage,
                            answer.toString()
                        ).whenComplete { res: Void?, er: Throwable? ->
                            if (er != null) {
                                Logger.instance
                                    .e("Failed to send answer: ", er)
                                result.completeExceptionally(er)
                            }
                            println(answer.toString() + "CompletableFuture2")
                            result.complete(null)
                        }
                    } catch (ex: Exception) {
                        result.completeExceptionally(ex)
                    }
                }
        } catch (ex: Exception) {
            result.completeExceptionally(ex)
        }
        return result
    }
}